lib/import_export/base_exporter.rb

```diff
@@ -0,0 +1,161 @@
+module ImportExport
+  class BaseExporter
+
+    attr_reader :export_data, :categories, :topics
+
+    CATEGORY_ATTRS = [:id, :name, :color, :created_at, :user_id, :slug, :description, :text_color,
+                      :auto_close_hours, :parent_category_id, :auto_close_based_on_last_post,
+                      :topic_template, :suppress_from_homepage, :all_topics_wiki, :permissions_params]
+
+    GROUP_ATTRS = [ :id, :name, :created_at, :mentionable_level, :messageable_level, :visibility_level,
+                    :automatic_membership_email_domains, :automatic_membership_retroactive,
+                    :primary_group, :title, :grant_trust_level, :incoming_email]
+
+    USER_ATTRS = [:id, :email, :username, :name, :created_at, :trust_level, :active, :last_emailed_at]
+
+    TOPIC_ATTRS = [:id, :title, :created_at, :views, :category_id, :closed, :archived, :archetype]
+
+    POST_ATTRS = [:id, :user_id, :post_number, :raw, :created_at, :reply_to_post_number, :hidden,
+                  :hidden_reason_id, :wiki]
+
+    def categories
+      @categories ||= Category.all.to_a
+    end
+
+    def export_categories
+      data = []
+
+      categories.each do |cat|
+        data << CATEGORY_ATTRS.inject({}) { |h, a| h[a] = cat.send(a); h }
+      end
+
+      data
+    end
+
+    def export_categories!
+      @export_data[:categories] = export_categories
+
+      self
+    end
+
+    # export groups that are mentioned in category permissions
```

Comment:
Generally we don't add comments since the method name should give a general idea of that the method does.